File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.java
class Solution {
    int n;
    int[][] dp = new int[100001][3];

    public int binarySearch(List<int[]> events, int endTime) {
        int l = 0, r = n - 1, ans = n;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (events.get(mid)[0] > endTime) {
                ans = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return ans;
    }

    public int solve(List<int[]> events, int i, int count) {
        if (count == 2 || i >= n) {
            return 0;
        }
        if (dp[i][count] != -1) {
            return dp[i][count];
        }

        int nextValidEventIndex = binarySearch(events, events.get(i)[1]);
        int take = events.get(i)[2] + solve(events, nextValidEventIndex, count + 1);
        int notTake = solve(events, i + 1, count);
        return dp[i][count] = Math.max(take, notTake);
    }

    public int maxTwoEvents(int[][] events) {
        n = events.length;
        List<int[]> eventList = new ArrayList<>();
        for (int[] event : events) {
            eventList.add(event);
        }
        Collections.sort(eventList, (a, b) -> Integer.compare(a[0], b[0]));
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }
        return solve(eventList, 0, 0);
    }
}

File: 2054-two-best-non-overlapping-events/2054-two-best-non-overlapping-events.py
class Solution:
    def __init__(self):
        self.n = 0
        self.dp = [[-1] * 3 for _ in range(100001)]

    def binarySearch(self, events, endTime):
        l, r, ans = 0, self.n - 1, self.n
        while l <= r:
            mid = l + (r - l) // 2
            if events[mid][0] > endTime:
                ans = mid
                r = mid - 1
            else:
                l = mid + 1
        return ans

    def solve(self, events, i, count):
        if count == 2 or i >= self.n:
            return 0
        if self.dp[i][count] != -1:
            return self.dp[i][count]

        nextValidEventIndex = self.binarySearch(events, events[i][1])
        take = events[i][2] + self.solve(events, nextValidEventIndex, count + 1)
        notTake = self.solve(events, i + 1, count)
        self.dp[i][count] = max(take, notTake)
        return self.dp[i][count]

    def maxTwoEvents(self, events: List[List[int]]) -> int:
        self.n = len(events)
        events.sort()
        self.dp = [[-1] * 3 for _ in range(self.n)]
        return self.solve(events, 0, 0)